### PR TITLE
Add illiterate trait

### PR DESCRIPTION
**Pull request by Rainfey**
## About the PR
Adds the Illiterate trait, making it (mostly) impossible to communicate via paper, labels, mass media, crayons/decals and communications console
Also fixes Mass Media console access checks - now requires command or service access

## Why / Balance
RP
I saw mention of something like this in #23303
Possible uses in any role that shouldnt be able to communicate with others (monkeys, mimes, future content)
NewsSystem annoyed me when i looked at it

## Technical details
- New Events OnReadAttempt, OnWriteAttempt added to PaperSystem, HandLabelerSystem, CrayonSystem and NewsSystem
- Hooked into CommunicationConsoleAnnouncementEvent and allowed listeners to modify the message
- Both events are handled by IlliterateSystem
- If the reader/writer is illiterate then the text is replaced by random characters
- The text remains unmodified for literate reader if it was written by a literate writer
- Non player sources of text such as cargo bounties are also unreadable by illiterates.
- Non letter/number characters are unaffected, to allow for drawing on paper
- Unfortunately i couldn't scramble labels written by literate people - illiterate people can still read those.
- Communications Console announcements are handled.
- Crayons will randomise if a letter or number is chosen, required adding a new field to decals
- Since i was in NewsSystem, i updated it to use IdCardSystem and to check against the AccessReader for adding and deleting articles

## Media

https://github.com/space-wizards/space-station-14/assets/11758391/fcb7cea1-7186-493c-9181-772a0ef3b1c5

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- add: Illiterate trait

### CODE REVIEW

**Comment by brainfood1183** [comment]
fun when you are illiterate mime.

**Comment by Rainfey** [comment]
Indeed, cognazine code is smelly though.

**Comment by Aisu9** [comment]
doesn't this also modify any other reading based interfaces like cargo request console and crew monitoring console? also Command, Medical and Chemistry shouldn't be able to roll with this trait on them because the essence of these role are reading a lot of things

**Comment by Hmeister-fake** [comment]
Animals in general should have this, even when cogged imo.

**Comment by Rainfey** [comment]
It doesnt affect user interfaces, just paper, labels (partially), the communications console and crayons

**Comment by Hmeister-fake** [comment]
> It doesnt affect user interfaces, just paper, labels (partially), the communications console and crayons

does it effect Decals, like letter decals?

**Comment by Rainfey** [comment]
Yes (crayons), i dont like i have implemented it - but if you try to draw a letter or number with a crayon it will choose a random letter/number instead.

**Comment by Cyber-Finn** [comment]
This is awesome, great job Rainfey!
I remember seeing Jean-Luc complain about something similar to a punpun, would defs make immersion/RP better

File: Content.Server/Crayon/CrayonSystem.cs [hunk 1]
```diff
@@ -112,7 +115,23 @@ private void OnCrayonBoundUI(EntityUid uid, CrayonComponent component, CrayonSel
         if (!_prototypeManager.TryIndex<DecalPrototype>(args.State, out var prototype) || !prototype.Tags.Contains("crayon"))
             return;
 
-        component.SelectedState = args.State;
+        //If this is a digit or letter decal, randomise it
+        if (prototype.IsDigitOrLetter)
```
Comment (metalgearsloth) [inline Content.Server/Crayon/CrayonSystem.cs:119]: Alphanumeric?

File: Content.Server/Paper/PaperSystem.cs [hunk 2]
```diff
@@ -147,6 +150,14 @@ private void OnInputTextMessage(EntityUid uid, PaperComponent paperComp, PaperIn
             {
                 paperComp.Content = args.Text;
 
+                var ev = new WriteAttemptEvent(args.Session.AttachedEntity, uid);
+                RaiseLocalEvent<WriteAttemptEvent>(uid, ev);
```
Comment (metalgearsloth) [inline Content.Server/Paper/PaperSystem.cs:154]: 1. Don't need the generic.
2. These events should be record structs raised by-ref.

File: Content.Server/Traits/Assorted/IlliterateSystem.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    [Dependency] private readonly IRobustRandom _random = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<WriteAttemptEvent>(OnWriteAttempt);
+        SubscribeLocalEvent<ReadAttemptEvent>(OnReadAttempt);
+        SubscribeLocalEvent<CommunicationConsoleAnnouncementEvent>(OnAnnouncement);
+    }
```
Comment (metalgearsloth) [inline Content.Server/Traits/Assorted/IlliterateSystem.cs:26]: There's no reason for illiteratesystem to be subscribing to its own event, you should just have callers call illiteratesystem directly instead as it's unnecessary overhead.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by metalgearsloth** [comment]
I ended up turning it into a review when there was more than I thought don't @ me

File: Content.Server/Crayon/CrayonSystem.cs (hunk 1 above)
Comment (Rainfey) [reply Content.Server/Crayon/CrayonSystem.cs]: Renamed

File: Content.Server/Paper/PaperSystem.cs (hunk 2 above)
Comment (Rainfey) [reply Content.Server/Paper/PaperSystem.cs]: Removed the events, callers use the method directly now (see comment on IlliterateSystem)

File: Content.Server/Traits/Assorted/IlliterateSystem.cs (hunk 3 above)
Comment (Rainfey) [reply Content.Server/Traits/Assorted/IlliterateSystem.cs]: Removed, callers use the method directly

File: Content.Shared/Decals/DecalPrototype.cs [hunk 4]
```diff
@@ -33,5 +33,11 @@ public sealed partial class DecalPrototype : IPrototype
         /// </summary>
         [DataField]
         public bool DefaultSnap = true;
+
+        /// <summary>
+        /// True if this decal represents a digit or letter, for illiteracy
+        /// </summary>
+        [DataField("isDigitOrLetter")]
+        public bool IsDigitOrLetter = false;
     }
```
[the remark this answers is not in the document]
Comment (Rainfey) [reply Content.Shared/Decals/DecalPrototype.cs]: Fixed, renamed and removed the string

File: Content.Shared/Traits/Assorted/SharedIlliterateSystem.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    /// What are they attempting to read
+    /// </summary>
+    public EntityUid? EntityRead;
+
+    public ReadAttemptEvent(EntityUid? reader, EntityUid? entityRead)
+    {
+        Reader = reader;
+        EntityRead = entityRead;
+        CanRead = true;
+    }
```
[the remark this answers is not in the document]
Comment (Rainfey) [reply Content.Shared/Traits/Assorted/SharedIlliterateSystem.cs]: I removed the event, unneeded.

**Pull request closed without merging**